Bottom: 383688f29e05da0a2381b972d1d05986b6755b34
Top:    c39169bcd76aa96717da6200420c3031aaf520e6
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-03-05 16:23:54 +0000

Compute the correct boundingbox for a polygon with arc segments.


---

diff --git a/src/misc.c b/src/misc.c
index 6ef9074..3fd9dc6 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -430,6 +430,27 @@ SetPolygonBoundingBox (PolygonType *Polygon)
     MAKEMIN (Polygon->BoundingBox.Y1, point->Y);
     MAKEMAX (Polygon->BoundingBox.X2, point->X);
     MAKEMAX (Polygon->BoundingBox.Y2, point->Y);
+
+    if (point->included_angle != 0)
+      {
+        BoxType arc_bound;
+        Coord cx, cy, radius;
+        Angle start_angle, delta_angle;
+        PointType *next_point;
+
+        /* NB: This next line uses the polygon point index, n, defined by POLYGON_POINT_LOOP */
+        next_point = &Polygon->Points[next_contour_point (Polygon, n)];
+
+        calc_arc_from_points_and_included_angle (point, next_point, point->included_angle,
+                                                 &cx, &cy, &radius, &start_angle, &delta_angle);
+
+        arc_bound = calc_thin_arc_bounds (cx, cy, radius, radius, start_angle, delta_angle);
+
+        MAKEMIN (Polygon->BoundingBox.X1, arc_bound.X1);
+        MAKEMIN (Polygon->BoundingBox.Y1, arc_bound.Y1);
+        MAKEMAX (Polygon->BoundingBox.X2, arc_bound.X2);
+        MAKEMAX (Polygon->BoundingBox.Y2, arc_bound.Y2);
+      }
   }
   /* boxes don't include the lower right corner */
   close_box(&Polygon->BoundingBox);
